### ISSUE TITLE
SetUp/TearDown methods are invoked multiple times before/after test in .NET Standard targeted projects

### ISSUE THREAD

**Issue by jalbrzym** [nunit/nunit#2626]
When I run test that is in test fixture that **inherits** from class containing **virtual method** in base class marked as [SetUp]/[TearDown] AND other derived classes overrides this method, then method **[SetUp] is invoked multiple times** before test execution AND method marked as **[TearDown] is invoked multiple times** after test execution - as many times as method is overriden throughout chain of inheritance. 

This issue occurs **only in .NET Standard** targeted projects.

Platform:
- NUnit 3.9.0
- NUnit 3 TestAdapter 3.9.0
- Visual Studio 2017 (v. 15.5.2)
- .NET Core SDK 2.1.2
- .NET Standard 1.6 targeted project

Sample:

```csharp
[TestFixture]
public class SampleIntegrationTest : SqlServerDatabaseIntegrationTest
{
    private int _setupCounter = 0;

    [Test]
    public void SetUpShouldInvokeOnce()
    {
        Assert.AreEqual(1, _setupCounter);
    }

    // SetUp method runs 3 times before test
    public override void SetUp()
    {
        ++_setupCounter;

        base.SetUp();
        // Additional setup logic
    }
}

public abstract class SqlServerDatabaseIntegrationTest : DatabaseIntegrationTest
{
    public override void SetUp()
    {
        base.SetUp();
        // Additional setup logic
    }
}

public abstract class DatabaseIntegrationTest : BaseIntegrationTest
{
    public override void SetUp()
    {
        base.SetUp();
        // Additional setup logic
    }
}

public abstract class BaseIntegrationTest
{
    [SetUp]
    public virtual void SetUp()
    {
    }
}
```

**Comment by jalbrzym** [comment]
After quick investigation I found that there is a problem with GetMethods extension method: 
[https://github.com/nunit/nunit/blob/04c6b7b4984a01831be94baeab0744b1cc686980/src/NUnitFramework/framework/Internal/Reflect.cs#L78](url)

.NET Standard targeted project uses custom extension method:
https://github.com/nunit/nunit/blob/3c028034e717c79bf00661c5852b358f9097d2e1/src/NUnitFramework/framework/Compatibility/ReflectionExtensions.cs#L332

As opposed to GetMethods standard implementation in Type class, it returns all methods throughout inheritance chain. This causes a polymorphic multiple call to the SetUp/TearDown method.

**Comment by ChrisMaddock** [comment]
This definitely sounds feasible. We have custom extension methods for the .NET Standard builds to replicate missing functionality that we use in our .NET Framework builds.

We'd appreciate a pull request here, if you'd be interested in digging further?

**Comment by jalbrzym** [comment]
@ChrisMaddock 
Sure, I'll propose solution of this issue and create pull request here.

**Comment by ChrisMaddock** [comment]
Great - look forward to seeing it! 😄

**Comment by mikkelbu** [comment]
This issue was fixed by #2626, but not auto-closed.
